### PR TITLE
quic: add QUIC connection close stats to upstream connections

### PR DESCRIPTION
**Pull request by RenjieTang**
Commit Message: Allow upstream quic connections to charge connection close stats.
Additional Description: For upstream connections, QuicStatNames reference is now passed down as Server->ClusterManager->ConnPool->EnvoyQuicConnection. 
Risk Level: Low
Testing: unit tests
Docs Changes: docs/root/configuration/upstream/cluster_manager/cluster_stats.rst

### CODE REVIEW

File: source/common/http/http3/conn_pool.cc
Comment (jmarantz) [inline source/common/http/http3/conn_pool.cc:66]: I think for non-test uses I'd avoid capturing `[&]` -- can you capture just what's needed?

File: test/common/http/conn_pool_grid_test.cc
Comment (jmarantz) [inline test/common/http/conn_pool_grid_test.cc:140]: I think typically in tests if we are instantiating a store we'd call the member variable `stats_` or `store_` or `stats_store_`.

Here & below.

File: test/integration/server.h
Comment (jmarantz) [inline test/integration/server.h:608]: use `OptRef<Quic:QuicStatNames>` rather than a pointer here. It's more idiomatic for Envoy style, and also avoids having an uninitialized value there potentially.

File: test/integration/utility.cc
Comment (jmarantz) [inline test/integration/utility.cc:235]: use `UNREFERENCED_PARAMETER(quic_stat_names)` in ./source/common/common/macros.h

File: test/integration/utility.cc
Comment (jmarantz) [inline test/integration/utility.cc:193]: given the existing usage model and implementation of this, where mock_stats_store is instantiated a few lines up, I think it would be more consistent for you to instantiate a new `Quic::QuicStatNames` here, using `mock_stats_store.symbolTable()`.

That will avoid the new args to makeSingleRequest, and remove a large number of other changes in this PR I think.

Comment (RenjieTang) [reply test/integration/utility.cc:193]: That's an great advice!
done.

File: source/common/upstream/cluster_manager_impl.h
Comment (jmarantz) [inline source/common/upstream/cluster_manager_impl.h:56]: WDYT of replacing the QuicStatNames args here and elsewhere in this PR with a new class Quic::Context. This would parallel the Http::Context and Grpc::Context names above, and would contain (own) the QuicStatNames. This would enable you to easily plumb in other things you might need globally in Quic.

I'm thinking in particular of a new Descriptor object that @RyanTheOptimist might want to add in another PR. If that never happens it's OK, but that way you'd be laying conduit that could be used for other cables in the future.

Thanks!

**Review by jmarantz — COMMENTED** [review]
this basically looks great, with a couple of small nits and a suggestion.

File: source/common/upstream/cluster_manager_impl.h
Comment (RenjieTang) [reply source/common/upstream/cluster_manager_impl.h:56]: I chatted with Ryan offline. He has no plan to land the CounterGroup change in the near future. So for now I think we can skip introducing the QuicContext wrapper.
If in a few months he decides to work on that change again, I can do a refactor to make the conduit.

**Review by jmarantz — APPROVED** [review]
OK -- I was worried that the lack of a Quic::Context might be a barrier to easily finishing off that PR, and also favor the parallel with the Http::Context and Grpc::Context.

But I don't feel that strongly about it, and you can always add it later (albeit with another large PR).

/assign-from @envoyproxy/senior-maintainers

**Comment by alyssawilk** [comment]
want to check out CI and I can take a look once that's sorted?
/wait

File: test/common/quic/envoy_quic_client_session_test.cc
Comment (alyssawilk) [inline test/common/quic/envoy_quic_client_session_test.cc:291]: do we have docs for this?
I'm not entirely sure where we'd put them.  Currently the upstream stats I'm familiar with are rooted in the cluster so docced in cluster_stats.rst
That said these are only updated when traffic is sent to a cluster, so maybe put them there but make it clear they're not rooted in the cluster and are in fact global?  Unless @mattklein123 or @phlax have better suggestions

**Review by alyssawilk — COMMENTED** [review]
Nice, thanks for following up on this!

File: test/common/quic/envoy_quic_client_session_test.cc
Comment (RenjieTang) [reply test/common/quic/envoy_quic_client_session_test.cc:291]: Yeah I think that might be the best place to put this doc. done.

**Review by alyssawilk — DISMISSED** [review]
LGTM, though it looks like one comment of @jmarantz may not be addressed - I'll let him check and merge.
/retest for the windows flake

File: test/mocks/server/BUILD
Comment (RenjieTang) [reply test/mocks/server/BUILD:185]: test/mocks/server/instance.h uses QuicStatNames, so it needs to be kept.

File: test/integration/utility.h
Comment (jmarantz) [reply test/integration/utility.h:18]: it shouldn't be necessary to include quic_stat_names.h here in the header; you should include it from the cc

File: test/mocks/server/BUILD
Comment (jmarantz) [reply test/mocks/server/BUILD:185]: I think I was wrong with this suggestion; this will be needed. Resolving.

File: test/integration/utility.h
Comment (RenjieTang) [reply test/integration/utility.h]: done including it in the cc file

**Review by jmarantz — DISMISSED** [review]

**Comment by alyssawilk** [comment]
unfortunately needs a main merge before we can land this

**Comment by alyssawilk** [comment]
Hey Renjie, is the windows release failure that uint64 optional issue you mentioned?  Looks like it might need one more pass?

**Comment by RenjieTang** [comment]
> Hey Renjie, is the windows release failure that uint64 optional issue you mentioned? Looks like it might need one more pass?

Yes it's still the optional<uint64_t> issue. I've filed https://github.com/envoyproxy/envoy/issues/17409

**Review by jmarantz — APPROVED** [review]

**Pull request merged**
